database/spanner_test.go

```diff
@@ -69,3 +74,19 @@ func Test_MigrateAndRun(t *testing.T) {
 	defer rows.Close()
 	require.NoError(t, rows.Err())
 }
+
+func TestSpannerUniqueViolation(t *testing.T) {
+	errMsg := "Failed to insert row with primary key ({pk#primary_key:\"282f6ffcd9ba5b029afbf2b739ee826e22d9df3b\"}) due to previously existing row"
+	// Test backwards-compatible parsing of spanner.Error (soon to be deprecated) from Spanner client
+	oldSpannerErr := spanner.ToSpannerError(status.New(codes.AlreadyExists, errMsg).Err())
+	if !database.SpannerUniqueViolation(oldSpannerErr) {
+		t.Error("should have matched unique violation")
+	}
+
+	// Test new apirerror.APIError response from Spanner client
+	newSpannerErr, parseErr := apierror.FromError(status.New(codes.AlreadyExists, errMsg).Err())
+	require.True(t, parseErr)
+	if !database.SpannerUniqueViolation(newSpannerErr) {
+		t.Error("should have matched unique violation")
+	}
```

Comment:
@gyao852 could you add one more test case checking the unique violation is still caught if we wrap the error with `fmt.Errorf("wrapped err: %w")`

Reply:
Ah, so if we wrap the error, the spanner library doesn't unwrap it since it ideally assumes it should be one of the two client responses above: 
![Screenshot 2023-05-03 at 6 01 16 PM](https://user-images.githubusercontent.com/28032700/236059790-1ccfeea5-1431-4ea9-bda8-207febd2340f.png)

When would we expect the error to be wrapped? Because I thought the repo layer will first catch this, and then we flag it ourselves before passing that upstream?

Regardless I re-added the `match` statement since I **think** all `err.Error()` will contain the base description and thus, Google's `Failed to insert row with primary key` string:
![Screenshot 2023-05-03 at 6 05 06 PM](https://user-images.githubusercontent.com/28032700/236060130-66dc89c5-8200-471e-a5cb-b6b69e42bade.png)

Reply:
Oh, looks like this is still an open issue: https://github.com/grpc/grpc-go/issues/2934